edx_prefectutils/amplitude.py

```diff
@@ -0,0 +1,347 @@
+        # thread.  Also, it is important that we delete that slice from the accumulated_events so that the python
+        # garbage collector will free all that memory as soon as the corresponding worker thread returns.
+        events_for_next_worker = accumulated_events[:last_idx_for_worker]
+        del accumulated_events[:last_idx_for_worker]
+
+        # Finally, batch the events such that each batch is small enough to fit inside a single Amplitude API call.
+        yield batch_generator(events_for_next_worker, batch_size)
+
+
+def _get_old_high_watermark(sf_credentials: SFCredentials, sf_role: str):
+    """
+    Get the amplitude event ID of the event immediately following the very last event sent to Amplitude's API.
+
+    Raises:
+        snowflake.connector.ProgrammingError: if the loader log table is inaccessible for some reason.
+    """
+    sf_connection = create_snowflake_connection(sf_credentials, sf_role)
+    query = """
+        select max(amplitude_event_id)
+          from prod.amplitude_events.amplitude_events_loader_log
+        """
+    cursor.execute(query)
+    old_high_watermark = cursor.fetchone()[0]
+    if not old_high_watermark:
+        old_high_watermark = 0
+
+
+def _get_new_high_watermark(sf_credentials: SFCredentials, sf_role: str):
+    """
+    TODO: make this more robust (e.g. what if the table does not exist?  what if the table is empty?)
+    """
+    sf_connection = create_snowflake_connection(sf_credentials, sf_role)
+    cursor = sf_connection.cursor()
+    # Get the current maximum amplitude_event_id.
+    query = """
+        select max(amplitude_event_id)
+          from prod.amplitude_events.amplitude_events
+        """
+    cursor.execute(query)
+    return cursor.fetchone()[0]
+
+
+def get_processing_segments():
+    # Get the old (i.e. current) high watermark.
+    old_high_watermark = _get_old_high_watermark()
+
+    # Get the new high watermark, i.e. the current maximum amplitude_event_id.
+    new_high_watermark = _get_new_high_watermark(sf_credentials, sf_role)
+
+    # Get a list of processing segments which cover all events between the old and
+    # new high watermark. The result is a list of start/end tuples describing
+    # amplitude_event_id ranges, where each range represents a single processing
+    # segment with the configured batch size.  The start ID is inclusive, whereas
+    # the end ID is exclusive.
+    #
+    # e.g. if old=1000000, new=3600000, bucketsize=1000000, then:
+    # processing_segments = [
+    #     (1000000, 2000000),
+    #     (2000000, 3000000),
+    #     (3000000, 3600001),
+    # ]
+    #
+    # In the above example, the last batch is smaller than 1m events because there
+    # are no more events to include.
+    return [
+        (processing_segment_start, min(processing_segment_start + processing_bucket_size, new_high_watermark + 1))
+        for processing_segment_start in range(old_high_watermark, new_high_watermark, processing_bucket_size)
```

Comment:
So I think this should start at old_high_watermark+1, and the min would be `min(processing_segment_start + processing_bucket_size - 1, new_high_watermark)`

Reply:
Yea, I had a few off-by-one errors that I fixed in my local index, but ultimately I want all the indexing and ranges to be pythonic.  I.e. start is always inclusive, end is always exclusive, and watermarks are more a descriptor of the start of the next batch.  I think that is off by one compared to your suggestion, if I understand correctly.  Take another look at these ranges and watermark definitions after I push up new changes.

Reply:
as discussed in a separate comment, I'll have to re-visit this to shift everything by one again.